src/main/java/com/aws/greengrass/deployment/DeploymentService.java

```diff
@@ -437,35 +440,55 @@ private void createNewDeployment(Deployment deployment) {
                                                                  JobStatus.IN_PROGRESS.toString(), new HashMap<>());
 
-        if (DEFAULT.equals(deployment.getDeploymentStage())
-                && DeploymentType.LOCAL.equals(deployment.getDeploymentType())) {
-            try {
-                copyRecipesAndArtifacts(deployment);
-            } catch (InvalidRequestException | IOException e) {
-                logger.atError().log("Error copying recipes and artifacts", e);
-                HashMap<String, String> statusDetails = new HashMap<>();
-                statusDetails.put("error", e.getMessage());
-                deploymentStatusKeeper.persistAndPublishDeploymentStatus(deployment.getId(),
-                        deployment.getDeploymentType(), JobStatus.FAILED.toString(), statusDetails);
-                return;
+        if (DEFAULT.equals(deployment.getDeploymentStage())) {
+
+            List<String> requiredCapabilities = deployment.getDeploymentDocumentObj().getRequiredCapabilities();
+            if (requiredCapabilities != null && !requiredCapabilities.isEmpty()) {
+                List<String> missingCapabilities = requiredCapabilities.stream()
+                        .filter(reqCapabilities -> !kernel.getSupportedCapabilities().contains(reqCapabilities))
+                        .collect(Collectors.toList());
+                if (!missingCapabilities.isEmpty()) {
+                    DeploymentResult result = new DeploymentResult(DeploymentStatus.FAILED_NO_STATE_CHANGE,
+                            new MissingRequiredCapabilitiesException("Missing required capabilities: "
+                                    + String.join(", ", missingCapabilities)));
+                    CompletableFuture<DeploymentResult> process = CompletableFuture.completedFuture(result);
+                    currentDeploymentTaskMetadata = new DeploymentTaskMetadata(deploymentTask, process,
+                            deployment.getId(), deployment.getDeploymentType(), new AtomicInteger(1),
+                            deployment.getDeploymentDocumentObj(), false);
+                    return;
+                }
             }
-        }
 
-        try {
-            if (DEFAULT.equals(deployment.getDeploymentStage())) {
+            if (DeploymentType.LOCAL.equals(deployment.getDeploymentType())) {
+                try {
+                    copyRecipesAndArtifacts(deployment);
+                } catch (InvalidRequestException | IOException e) {
+                    logger.atError().log("Error copying recipes and artifacts", e);
+                    HashMap<String, String> statusDetails = new HashMap<>();
+                    statusDetails.put("error", e.getMessage());
+                    deploymentStatusKeeper.persistAndPublishDeploymentStatus(deployment.getId(),
+                            deployment.getDeploymentType(), JobStatus.FAILED.toString(), statusDetails);
```

Comment:
Why is this error handling so different to that on L451-L458?

Reply:
Both does similar things, refactored to use the same logic to mark the deployment as failed.